### ISSUE TITLE
Add a pessimistic Backtrace type

### ISSUE THREAD

**Issue by shepmaster** [shepmaster/snafu#88]
It's hard to tell if a foreign error type has a backtrace or not. If it does, we don't want to create our own, otherwise we do.

It could be interesting to try and have some type that checks if the underlying error has a backtrace. If so, it does nothing. Otherwise, it generates one at that point.

**Comment by Lej77** [comment]
I guess this would be accomplished with the derive macro. I have written a macro that can be used on a type and if the type implements a certain trait then it can do something with that trait. 

[Here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=aa7c575ce5caee323f1f38ffd57dbb1b) is all the code I wrote when testing it out how this could work. It became quite a lot of code so [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5bc2b947e4d4c61b902ad13371edc0c) is a simplified version that includes as little unnecessary code as possible.

# Code explanation

The code works using "resolver" structs that has a method that takes an error and returns its backtrace. The structs contain a `PhantomData` that is used to store the error type. This means that the structs' impl blocks can have where clauses based on the error type. This is used to only implement the method that gets the backtrace if the error type actually has a backtrace. 

The code uses the `std::ops::Deref` trait to allow for a fallback method. If the method that gets the backtrace isn't implemented for the resolver then it derefs to a struct that has a method with the same signature but that just returns `None`. 

The macro has the constraint that it requires that no conflicting traits are in scope. Conflicting traits are traits that are implemented for the resolver and that provides a method with the same name that is used for the method that gets the backtrace from an error. I can see two ways to work around that issue:
1. Use a module in the macro to get a new scope without any user defined traits. This is the solution that my example code uses.
2. Use a method name that is really weird so that it is very unlikely that the user will have a trait in scope that has the same method name.

**Comment by shepmaster** [comment]
Wow, that's a lot of hard work (and code)! Maybe I was a bit optimistic that this would be an "easy" addition!

I was thinking something along the lines of [this](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=f91b72bd56e4983ba5d1a219638481da). 

TL;DR, we do need some help from the derive macro to appropriately call `PessimisticBacktrace::new` with a reference to the error and to call `pbwo.0.as_ref().or_else(|| wo.backtrace()` at the appropriate time. This is ugly because it makes that type "special ", although we could introduce traits to make it a bit more general if it was useful.

What do you think?

**Comment by Lej77** [comment]
Your code assumes that the source error implements the `ErrorCompat` trait while my code doesn't. This means that the macro could only generate your code if the `ErrorCompat` trait is implemented for the source error otherwise it would cause a compile time error. If this trait is not already required (which I don't think it is) then the user would need to annotate their source error with an attribute or something to tell the macro to generate your code.

One way to be able to always generate the source backtrace checking code is if specialization is used to allow a blanket implementation of `ErrorCompat` for errors that don't manually implement it. Since specialization is unstable this isn't currently possible (at lest on stable rust). 

My code is a hack for specialization that allows the code to work even if the source error doesn't implement `ErrorCompat`. This means that the code could always be generated without the user annotating the source error with an attribute or something like that.
